### ISSUE TITLE
Quoted identifiers in binary encoding

### ISSUE THREAD

**Issue by Nadrieril** [dhall-lang/dhall-lang#462]
```
$ echo '`Bool`' | dhall type
Error: Unbound variable: Bool
$ echo '`Bool`' | dhall encode | dhall decode | dhall type
Type
```
The standard specifies to encode all variables with index zero as strings, but this can clash with the encoding of builtins.

**Comment by Nadrieril** [comment]
Just for fun:
```
$ echo '(\(`Natural/even`: Natural -> Bool) -> `Natural/even`) Natural/odd 0' | dhall normalize
False
$ echo '(\(`Natural/even`: Natural -> Bool) -> `Natural/even`) Natural/odd 0' | dhall encode | dhall decode | dhall normalize
True
```

**Comment by Gabriel439** [comment]
@Nadrieril: My reading of the grammar is that:

```haskell
`Bool`
```

... should be treated the same as:

```haskell
Bool
```

... and therefore this is an issue with the Haskell implementation rather than an issue with the standard.

**Comment by Gabriel439** [comment]
Whoops, I take that back.  I completely forgot that the entire point of quoting the identifiers was to avoid conflicting with built-ins 🙂

**Comment by singpolyma** [comment]
> Whoops, I take that back. I completely forgot that the entire point of quoting the identifiers was to avoid conflicting with built-ins 🙂

I'm curious: what's the use case for allowing variables named `Bool`, &c?

Record keys I understand, but variables are an internal matter.‎

**Comment by Gabriel439** [comment]
@singpolyma: Maybe you just really want to name a variable `Type` 🙂 , like in this example: https://github.com/dhall-lang/dhall-lang/issues/469#issuecomment-480664515

**Comment by philandstuff** [comment]
I think we have a problem here, which is that after #399 there are multiple reasonable interpretations of how to map identifiers to builtins.

I have only recently run `git pull` to see the latest ABNF and semantics.md, and I have largely not been following these discussions.  There is nowhere in the standard that explicitly communicates how the ABNF `identifier` rule maps on to the variables and builtins in semantics.md.  I think there are at least two reasonable interpretations:

1. the literal identifiers `Bool`, `Natural` etc get parsed directly into builtin values
2. the literal identifiers `Bool`, `Natural` etc are merely special bindings of variables, and subject to the same shift, subst judgements as other variables.

It seems that I have, in my independent reading of the ABNF and semantics, naturally come to interpretation 2.  The fact that @Gabriel439 temporarily accepted this in https://github.com/dhall-lang/dhall-lang/issues/462#issuecomment-480655044 shows that this is a reasonable position. However, it seems like the rest of the community is behind interpretation 1.

I want to break down each interpretation a little bit, talk about the pros and cons, before proposing that we come to some sort of decision.

## `Bool` is a special identifier

In this interpretation, whenever we see an identifier with a builtin name like `Bool`, this is parsed directly into the builtin.  This means that the identifier `Bool` cannot be bound by `let` or lambda or pi.

### Advantages

There is circumstantial evidence that this was the intention - in particular, the ban on bindings which capture builtin names.

### Disadvantages

This interpretation requires extra parsing rules which aren't captured in the grammar. I feel strongly that the grammar should capture what it means to parse dhall. If there are special rules around special identifiers that aren't captured in the grammar, those special rules ought to be at least captured somewhere else.  And if we need to write them down anyway, why not do so in the grammar? (In other words, I think if we accept this interpretation, we should revert #399).

Under this interpretation, there's a clash between the CBOR representation of `` `Bool` `` (a variable with name `Bool`) and `Bool` (the Bool builtin). (This is the problem that #488 has been raised to fix).

### Questions

Under this interpretation, what is the meaning of an expression like `Bool@1`? This is a valid parse according to the ABNF, but it's not super clear what it could mean.  I suppose the most likely answer is that it's the same as `` `Bool`@1``. Personally, it's confusing that `Bool` is not the same as `` `Bool` `` but `Bool@1` is the same as `` `Bool`@1 ``.

## `Bool` is a variable binding

Under this interpretation, the Bool builtin is just the defined behaviour of a free variable with the special name `Bool` (and de Bruijn index 0).  Because builtins are just a subset of all variables, they are affected by the shift and subst judgements, and can be bound.  This means that something like:

```
let `Bool` = Natural
in { foo = True, bar = 5 } : { foo : Bool@1, bar : Bool }
```
is perfectly valid.  We have to quote `Bool` in the binding because of the grammar rules, but in every other respect `` `Bool` `` is just the same as `Bool`.

### Advantages

We can keep a simple grammar with a single `identifier` rule and no special-case builtin rules. In particular, there isn't a disconnect between the ABNF and the AST in actual implementations. I feel that it's important that the ABNF captures the actual AST. More precisely, I think each type of AST node should map in a one-to-one way to an ABNF rule.

This matches neatly with the current CBOR encoding, so #488 isn't necessary.

### Disadvantages

The inability to directly bind builtin identifiers like `Bool` feels inconsistent under this interpretation. But we could always change that.

## Summary

It's pretty clear that I still prefer interpretation 2 (builtins are a subset of variables).  But even more important is that we just pick one!  Right now, I think both are reasonable interpretations according to the text of the standard.  We should pick one and make it explicit.

**Comment by Nadrieril** [comment]
Something similar [has been discussed](https://github.com/dhall-lang/dhall-lang/issues/314) previously.
To sum up the discussion there:

## `Bool` is a variable binding

### Advantages

- [we'd standardize "builtins extensibility" (e.g. the Clojure compiler doesn't have any extensibility point so far to define your own builtins, and it would be nice to have)](https://github.com/dhall-lang/dhall-lang/issues/314#issuecomment-445468966)
- ["builtins extensibility" wouldn't break standard-compliance nor existing code](https://github.com/dhall-lang/dhall-lang/issues/314#issuecomment-468783336)

### Disadvantages

- [it would complicate the standard](https://github.com/dhall-lang/dhall-lang/issues/314#issuecomment-445352353)
- [builtin shadowing seems to have no use-case](https://github.com/dhall-lang/dhall-lang/issues/314#issuecomment-445468966)
- [builtin shadowing could be confusing, in particular across imports](https://github.com/dhall-lang/dhall-lang/issues/314#issuecomment-445468966)

I am personally in favor of "`Bool` is a variable binding", mainly because I tend to prioritize language uniformity.

**Comment by f-f** [comment]
@philandstuff I think the "right interpretation" became somehow less clear after #399, so I agree that we should specify this better.
I think we're touching on some different topics here, so I'll answer separately:

#### About reverting #399

I see #399 mainly as a _cosmetic change_ to the grammar. In fact, I don't think the CBOR encoding is relevant here: the grammar specification and the CBOR specification are _orthogonal_ and are not supposed to be tied in any way except for the common language that we use to describe them.
And this is because their purposes are fairly disjoint:
- the grammar should _unambiguously_ describe the set of allowed expressions. This means that `Bool@2` being allowed is indeed a bug because it doesn't make sense in our current interpretation.
- the CBOR encoding should specify the _bijection_ that ties a Dhall expression to its binary representation. This means that this issue is indeed a bug - because it shows that this is actually not a bijection as it should be - and should be fixed (I didn't look at #488 yet tough, so I cannot say on that)

I'll also note that we purposefully don't want to standardize a canonical AST. The reason is well-explained by Gabriel here: https://github.com/dhall-lang/dhall-haskell/issues/568#issuecomment-419733127

So I don't think we should revert it even if we are in "interpretation 1", but I'll open a PR now to add a rule that implementors can use if they want to detect builtins in the grammar instead of doing it in a separate step

#### About using the grammar rules as AST

As I already mentioned in https://github.com/dhall-lang/dhall-lang/issues/394#issuecomment-468660235 at first I was also on a quest to have the grammar handle all aspects of parsing, that is: have the grammar rules map 1-1 to AST.
I had to give up because it's not possible to encode everything precisely enough in there, because ABNF is not powerful enough.
This means that we have to supplement with natural language in comments in the grammar. Before #399 we had rules for parsing builtins into, but that was ambiguous, so now we say "use identifiers and match for builtins".
To me it doesn't really matter, as long as it's specified in some way and decent enough to work with.

All the issues I had about this went away as soon as I defined a better AST representation in my implementation, less tied to the grammar 🙂 
In fact, "adding more intermediate representations" seems to be a good thing to do most of the time (e.g. I think a good demonstration of this is https://github.com/dhall-lang/dhall-haskell/pull/876, where cleverly adding another intermediate representation results in huge performance improvements)

#### About going for "interpretation 2" and treating builtins as bindings

I think there are good reasons for not treating builtins as normal variable bindings, and I stated them in https://github.com/dhall-lang/dhall-lang/issues/314#issuecomment-445468966 and https://github.com/dhall-lang/dhall-lang/issues/394#issuecomment-468660235.
Here I'm going to add that I'm not sure anymore that "standardizing builtins extensibility" (that I first listed as an advantage of this approach as @Nadrieril noted above) is a good thing, since it goes more towards helping the usecase of "Dhall as a general base for DSLs" than the main usecase we are targeting that is "displacing YAML"

**Comment by singpolyma** [comment]
> This means that `Bool@2` being allowed is indeed a bug because it doesn't make sense in our current interpretation.

I don't think Bool@2 needs to be a syntax error, though. In fact, it even has an obvious semantics (being the same as `Bool`@2) or if we don't like that then it would fail typechecking as a free variable.

‎I'm personally opposed to allowing bound variables with names that match builtins (the quoting makes no difference to me really, since it does not exist internally) however I understand the standard currently allows it, which means either builtins or quoted variable bindings has to be handled as special to implement the standard -- with builtins being the "more different" one to me.

**Comment by f-f** [comment]
> I don't think Bool@2 needs to be a syntax error, though. In fact, it even has an obvious semantics (being the same as `Bool`@2) or if we don't like that then it would fail typechecking as a free variable.

Yeah I added an edit above to say that we could actually just ignore the index for builtins (and this what the Clojure implementation currently does), and that I'm fine with either failing or ignoring the index as long as we specify the behaviour.

**Comment by singpolyma** [comment]
>  I'm fine with either failing or ignoring the index as long as we specify the behaviour.

Or the presence of an index could make it the same as escaping.   `Bool`@0 and Bool@0 could be the same and then you only need quoting in binding position (or if you prefer writing `Bool` to Bool@0)‎

**Comment by philandstuff** [comment]
@f-f

> the grammar should unambiguously describe the set of allowed expressions. 

If the grammar is the source of truth about what a Dhall expression is, then something needs to tie that to the list of expressions in semantics.md.  Currently we have one list of allowed expressions (defined syntactically), and a similar-but-not-identical list of allowed expressions (defined semantically), and no formal connection from one to the other. This is the root of the ambiguity.

The discussion we’re having about the meaning of `Bool@0` is precisely about how we map syntactic expressions to semantic expressions. 

Once we make a decision, where in the standard should we write it down? If we don’t want to write it in the ABNF, does it belong in semantics.md? Do we need a third document?

It worked before #399 because the lists were similar enough that there was only one obvious interpretation.

——

On a different subject, I hadn’t considered imports, and I agree that they really complicate interpretation 2 above.

**Comment by f-f** [comment]
@philandstuff this mapping happened in the grammar file so far. E.g. we have a comment about builtins already:
https://github.com/dhall-lang/dhall-lang/blob/7c25a52db1dc38a93df85959c96cb7f24740560d/standard/dhall.abnf#L335-L337

I just opened #490 to add back the builtin rules (albeit unused) and specify what happens when parsing things like `Bool@0`

**Comment by Gabriel439** [comment]
@philandstuff: What if we were to use the CBOR representation to provide an unambiguous link between the grammar and the semantics?

In other words, for a rule like `nonreserved-label` in the ABNF we specify in a comment what CBOR encoding it corresponds to and then for the matching case in the semantics document we also tag it with the same CBOR encoding?

**Comment by philandstuff** [comment]
I don’t think the CBOR is enough. In order to disambiguate the two interpretations above, something needs to say how different kinds of `identifier` map to different kinds of semantics.md expression. The CBOR encoding can’t provide this information because it doesn’t talk about syntax. 

In particular, I anticipate a comment above the `identifier` rule saying something like:

“If the label is a simple-label and matches one of the builtin names, this is a builtin. It is a syntax error to specify a de Bruijn index in this case. Otherwise, this is a variable with name and index matching the label and index.”

This unambiguously says we are in interpretation 1 and rules out interpretation 2. 

(I feel that this encodes in a comment what is better expressed as syntax rules.)

I think the equivalent for interpretation 2 is to accept that all `indentifier`s parse as variables, but to update the semantics so that all of the rules for builtins apply to variables instead.

(This would let us get rid of all the shift and subst judgements for builtins, though as has been pointed out we would need to handle imports carefully.)

**Comment by Gabriel439** [comment]
@philandstuff: So in #488 the CBOR representation for built-ins and variables differs, so you can use the equivalent CBOR encoding to distinguish them more precisely
